r/tests/testthat/test-dplyr.R

```diff
@@ -989,19 +989,86 @@ test_that("sign()", {
   )
 })
 
-test_that("ceiling(), floor(), trunc()", {
+test_that("ceiling(), floor(), trunc(), round()", {
   df <- tibble(x = c(-1, -0.55, -0.5, -0.1, 0, 0.1, 0.5, 0.55, 1, NA, NaN))
 
   expect_dplyr_equal(
     input %>%
       mutate(
         c = ceiling(x),
         f = floor(x),
-        t = trunc(x)
+        t = trunc(x),
+        r = round(x)
       ) %>%
       collect(),
     df
   )
+
+  # with digits set to 1
+  expect_dplyr_equal(
+    input %>%
+      filter(x %% 0.5 == 0) %>%
+      mutate(r = round(x, 1)) %>%
+      collect(),
+    df
+  )
+
+  # with digits set to -1
+  expect_dplyr_equal(
+    input %>%
+      mutate(
+        rd = round(floor(x * 111), -1), # double
+        y = ifelse(is.nan(x), NA_integer_, x),
+        ri = round(as.integer(y * 111), -1) # integer (with the NaN removed)
+      ) %>%
+      collect(),
+    df
+  )
+
+  # round(x, -2) is equivalent to round_to_multiple(x, 100)
+  expect_equal(
+    Table$create(x = 1111.1) %>%
+      mutate(r = round(x, -2)) %>%
+      collect(),
+    Table$create(x = 1111.1) %>%
+      mutate(r = arrow_round_to_multiple(x, options = list(multiple = 100))) %>%
+      collect()
+  )
+
+  skip_on_os("windows") # float representation error might cause inconsistency
+
+  expect_dplyr_equal(
+    input %>%
+      mutate(r = round(x, 1)) %>%
+      collect(),
+    df
+  )
+
+  # verify that round mode HALF_TO_EVEN which is what the round() binding uses
+  # yields results consistent with R...
+  expect_equal(
+    as.vector(
+      call_function(
+        "round",
+        Array$create(df$x),
+        options = list(ndigits = 1L, round_mode = RoundMode$HALF_TO_EVEN)
+      )
+    ),
+    round(df$x, 1)
+  )
+  # ... but round mode HALF_TOWARDS_ZERO does not.
```

Comment:
Are these inconsistencies attributed to floating-point error propagation or true errors in the resulting calculation?

Reply:
Basically, how large are the difference? `all.equal()` has parameters for controlling the tolerance of "approximate equality". This tolerance should be adjusted relative to the magnitudes of the values...so I am curious if these inconsistencies are observed throughout different ranges of values.

Reply:
The R docs say:
>Note that for rounding off a 5, the IEC 60559 standard (see also ‘IEEE 754’) is expected to be used, ‘go to the even digit’. Therefore round(0.5) is 0 and round(-1.5) is -2. However, this is dependent on OS services and on representation error (since e.g. 0.15 is not represented exactly, the rounding rule applies to the represented number and not to the printed number, and so round(0.15, 1) could be either 0.1 or 0.2).

So R's rounding mode is equivalent to the `HALF_TO_EVEN` mode in Arrow. The expectation beginning on line 1049 tests that. The expectation beginning on line 1060 tests the test.

The default tolerance is sufficient to detect that but not detect floating-point error propagation.

These tests are skipped on CRAN to avoid possible failures on CRAN test machines that could hypothetically be caused by the dependency of R's rounding behavior on OS services.

Reply:
Should we include a (truncated) version of that explanation in a comment here? I see that's what it's doing — but I'm not sure I would have without this comment.

Reply:
@ianmcook Thanks for the clarification, I understand now the valid/invalid tests you set up. I previously thought that HALF_TOWARDS_ZERO was set for both approaches but errors were showing up. It was my mistake, I did not read carefully the comments.

Reply:
Thanks @edponce and @jonkeane! I expanded the comments per these suggestions in c3d5d04